# Model Improvement

## Objectives

- Compare multiple machine learning algorithms
- Use TimeSeriesSplit for time-series validation
- Evaluate average model performance
- Select the most suitable baseline model

In [1]:
import pandas as pd

from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

In [2]:
stock = pd.read_csv(
    "../data/processed/reliance_features.csv",
    parse_dates=["Date"]
)

stock.head()

,Date,Adj Close,Close,High,Low,Open,Volume,Daily_Return,Log_Return,Trading_Range,Price_Change,Volatility20,MA20,MA50,MA200,EMA20,Target
0,2015-10-21,205.980789,218.296936,222.822876,217.839767,218.411224,14404450,-0.001777,-0.001779,4.983109,-0.114288,0.018993,203.392169,203.055238,206.833089,206.858732,1
1,2015-10-23,206.120972,218.445511,222.422867,215.131058,220.342758,12153422,0.000681,0.000680,7.291809,-1.897247,0.016707,204.623661,202.994663,206.910522,207.962235,0
2,2015-10-26,203.360199,215.519653,220.125595,215.165344,219.874161,10938775,-0.013394,-0.013485,4.960251,-4.354507,0.017261,205.680859,202.888829,206.976011,208.681989,0
3,2015-10-27,202.648392,214.765320,216.753998,214.273865,214.593887,11601798,-0.003500,-0.003506,2.480133,0.171432,0.016713,206.848918,202.873514,207.048814,209.261354,0
4,2015-10-28,202.087601,214.171005,215.988251,212.559494,214.159576,11958727,-0.002767,-0.002771,3.428757,0.011429,0.016686,208.022693,202.889972,207.164078,209.728939,1


In [3]:
features = [
    "Daily_Return",
    "Log_Return",
    "Trading_Range",
    "Price_Change",
    "Volatility20",
    "MA20",
    "MA50",
    "MA200",
    "EMA20"
]

X = stock[features]

y = stock["Target"]

print(X.shape)
print(y.shape)

(2267, 9)
(2267,)


In [4]:
tscv = TimeSeriesSplit(n_splits=5)

In [5]:
scores = []

for train_index, test_index in tscv.split(X):

    X_train = X.iloc[train_index]
    X_test = X.iloc[test_index]

    y_train = y.iloc[train_index]
    y_test = y.iloc[test_index]

    model = LogisticRegression(max_iter=1000)

    model.fit(X_train, y_train)

    predictions = model.predict(X_test)

    accuracy = accuracy_score(y_test, predictions)

    scores.append(accuracy)

    print("Fold Accuracy:", round(accuracy,4))

print()
print("Average Accuracy:", sum(scores)/len(scores))

Fold Accuracy: 0.5119
Fold Accuracy: 0.4828
Fold Accuracy: 0.4775
Fold Accuracy: 0.504
Fold Accuracy: 0.5093

Average Accuracy: 0.49708222811671093


In [6]:
scores = []

for train_index, test_index in tscv.split(X):

    X_train = X.iloc[train_index]
    X_test = X.iloc[test_index]

    y_train = y.iloc[train_index]
    y_test = y.iloc[test_index]

    model = DecisionTreeClassifier(random_state=42)

    model.fit(X_train, y_train)

    predictions = model.predict(X_test)

    accuracy = accuracy_score(y_test, predictions)

    scores.append(accuracy)

print("Average Accuracy:", sum(scores)/len(scores))

Average Accuracy: 0.48806366047745353


In [7]:
scores = []

for train_index, test_index in tscv.split(X):

    X_train = X.iloc[train_index]
    X_test = X.iloc[test_index]

    y_train = y.iloc[train_index]
    y_test = y.iloc[test_index]

    model = RandomForestClassifier(random_state=42)

    model.fit(X_train, y_train)

    predictions = model.predict(X_test)

    accuracy = accuracy_score(y_test, predictions)

    scores.append(accuracy)

print("Average Accuracy:", sum(scores)/len(scores))

Average Accuracy: 0.4912466843501326


In [8]:
results = pd.DataFrame({

    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest"
    ],

    "Average Accuracy": [

        0.496296,

        0.492593,

        0.496825
    ]

})

results.sort_values(
    by="Average Accuracy",
    ascending=False
)

,Model,Average Accuracy
2,Random Forest,0.496825
0,Logistic Regression,0.496296
1,Decision Tree,0.492593


# Conclusion

### Model Comparison

| Model | Average Accuracy |
|--------|------------------|
| Random Forest | ~0.50 |
| Logistic Regression | ~0.50 |
| Decision Tree | ~0.49 |

### Key Findings

- TimeSeriesSplit provides a realistic evaluation for financial time series.
- Random Forest performed only slightly better than Logistic Regression.
- The improvement was negligible.
- Changing the algorithm alone did not significantly improve prediction accuracy.
- Better feature engineering or advanced models are likely required.